sky/serve/autoscalers.py

```diff
@@ -132,93 +141,105 @@ def collect_request_information(
-            self.upscale_counter = 0
-        elif num_replicas > self.max_replicas:
-            num_replicas_delta = self.max_replicas - num_replicas
-            self.downscale_counter = 0
-        elif target_num_replicas > num_replicas:
+        if not self.bootstrap_done:
+            self.bootstrap_done = True
+            return target_num_replicas
+        elif target_num_replicas > self.target_num_replicas:
             self.upscale_counter += 1
             self.downscale_counter = 0
             if self.upscale_counter >= self.scale_up_consecutive_periods:
                 self.upscale_counter = 0
-                num_replicas_delta = target_num_replicas - num_replicas
-        elif target_num_replicas < num_replicas:
+                return target_num_replicas
+        elif target_num_replicas < self.target_num_replicas:
             self.downscale_counter += 1
             self.upscale_counter = 0
             if self.downscale_counter >= self.scale_down_consecutive_periods:
                 self.downscale_counter = 0
-                num_replicas_delta = target_num_replicas - num_replicas
+                return target_num_replicas
         else:
             self.upscale_counter = self.downscale_counter = 0
-        logger.info(f'Upscale counter: {self.upscale_counter}/'
-                    f'{self.scale_up_consecutive_periods}. '
-                    f'Downscale counter: {self.downscale_counter}/'
-                    f'{self.scale_down_consecutive_periods}')
+        return self.target_num_replicas
 
-        if num_replicas_delta == 0:
+    def evaluate_scaling(
+        self,
+        replica_infos: List['replica_managers.ReplicaInfo'],
+    ) -> List[AutoscalerDecision]:
+
+        alive_replica_infos = [info for info in replica_infos if info.is_alive]
+        num_alive_replicas = len(alive_replica_infos)
+
+        self.target_num_replicas = self._get_desired_num_replicas()
+        logger.info(
+            f'Final target number of replicas: {self.target_num_replicas} '
+            f'Upscale counter: {self.upscale_counter}/'
+            f'{self.scale_up_consecutive_periods}, '
+            f'Downscale counter: {self.downscale_counter}/'
+            f'{self.scale_down_consecutive_periods} '
+            f'Number of alive replicas: {num_alive_replicas}')
+
+        scaling_options = []
+        all_replica_ids_to_scale_down: List[int] = []
+
+        def _get_replica_ids_to_scale_down(
+            status_order: List['serve_state.ReplicaStatus'],
+            num_limit: int,
+        ) -> List[int]:
+
+            def get_status_order(status):
+                if status in status_order:
+                    return status_order.index(status)
+                else:
+                    return len(status_order)
+
+            alive_replica_infos_sorted = sorted(
+                alive_replica_infos,
+                key=lambda info: get_status_order(info.status))
```

Comment:
```suggestion
            alive_replica_infos_sorted = sorted(
                alive_replica_infos,
                key=lambda info: status_order.index(info.status) if info.status in status_order else len(status_order))
```
nit: try to reduce defining too many nested functions (3 in this case).